langchain/src/document_loaders/web/notionapi.ts

```diff
@@ -127,143 +205,160 @@ export class NotionAPILoader extends BaseDocumentLoader {
       typeof page === "string"
-        ? [this.notionClient.pages.retrieve({ page_id: page }), page]
+        ? [
+            this.limiter.schedule(() =>
+              this.notionClient.pages.retrieve({ page_id: page })
+            ),
+            page,
+          ]
         : [page, page.id];
 
     const [pageDetails, pageBlocks] = await Promise.all([
       pageData,
-      getBlockChildren(this.notionClient, pageId, null),
+      this.limiter.schedule(() =>
+        getBlockChildren(this.notionClient, pageId, null)
+      ),
     ]);
 
-    if (!isFullPage(pageDetails)) return [];
-
-    const { mdBlocks, childDocuments } = await this.loadBlocksAndDocs(
-      pageBlocks
-    );
+    if (!isFullPage(pageDetails)) return;
 
+    const mdBlocks = await this.loadBlocks(pageBlocks);
     const mdStringObject = this.n2mClient.toMarkdownString(mdBlocks);
-
     const pageDocument = new Document({
       pageContent: mdStringObject.parent,
       metadata: this.parsePageDetails(pageDetails),
     });
 
-    return [pageDocument, ...childDocuments];
+    this.documents.push(pageDocument);
+    this.pageCompleted.push(pageId);
+    this.onDocumentLoaded(
+      this.documents.length,
+      this.pageQueueTotal,
+      pageDocument.metadata.properties.title,
+      this.rootTitle
+    );
   }
 
-  private async loadDatabase(id: string): Promise<Document[]> {
-    const documents: Document[] = [];
-
+  private async loadDatabase(id: string) {
     try {
       for await (const page of iteratePaginatedAPI(
         this.notionClient.databases.query,
         {
           database_id: id,
+          page_size: 50,
         }
       )) {
-        if (!isFullPage(page)) continue;
-
-        documents.push(...(await this.loadPage(page)));
+        this.addToQueue(page.id);
       }
     } catch (e) {
       console.log(e);
       // TODO: Catch and report api request errors
     }
-
-    return documents;
   }
 
   async load(): Promise<Document[]> {
-    const documents: Document[] = [];
-
-    switch (this.type) {
-      case "page":
-        documents.push(...(await this.loadPage(this.id)));
-        break;
-      case "database":
-        documents.push(...(await this.loadDatabase(this.id)));
-        break;
-      default:
+    const resPagePromise = this.notionClient.pages
+      .retrieve({ page_id: this.id })
+      .then((res) => {
+        this.addToQueue(this.id);
+        return res;
+      })
+      .catch((error: APIResponseError) => error);
+
+    const resDatabasePromise = this.notionClient.databases
+      .retrieve({ database_id: this.id })
+      .then(async (res) => {
+        await this.loadDatabase(this.id);
+        return res;
+      })
+      .catch((error: APIResponseError) => error);
+
+    const [resPage, resDatabase] = await Promise.all([
+      resPagePromise,
+      resDatabasePromise,
+    ]);
+    const errors = [resPage, resDatabase].filter(isErrorResponse);
+
+    if (errors.length === 2) {
```

Comment:
Exactly 2?

Reply:
Yes, both resPage and resDatabase must error to throw an error. This part of the code determines whether the notion ID is a page or database. Only when both are errors will it throw. 

The `errors` variable is set in the line above and filter the results for both page and database calls and includes the filter with the error type guard. One is always expected to error.

Reply:
Ah gotcha. Maybe add a comment there?

Reply:
I added a comment to clarify the error checking